eth/api_migration.go

```diff
@@ -0,0 +1,752 @@
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// Policy
+// LOCAL
+// 1. Local first
+// 2. If not found, and erigon configured, forward request
+// FORWARD
+// 1. If block number is earlier than configured, forward request
+// 2. Otherwise, use local
+
+var (
+	errInvalidBlockRange = errors.New("invalid block range params")
+)
+
+// MigrationRPCService holds the configuration for RPC migration
+type MigrationRPCService struct {
+	MigrationBlock uint64
+	ErigonClient   *rpc.Client
+}
+
+// NewMigrationRPCService creates a new migration configuration
+func NewMigrationRPCService(config *ethconfig.Config) (*MigrationRPCService, error) {
+	if config.XLayer.RpcMigration.MigrationBlock == nil || config.XLayer.RpcMigration.PPRPCUrl == "" {
+		return nil, nil // Migration not configured
+	}
+
+	timeout := config.XLayer.RpcMigration.PPRPCTimeout
+	if timeout == 0 {
+		timeout = 10 * time.Second
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	erigonClient, err := rpc.DialContext(ctx, config.XLayer.RpcMigration.PPRPCUrl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to erigon RPC: %w", err)
+	}
+
+	return &MigrationRPCService{
+		MigrationBlock: *config.XLayer.RpcMigration.MigrationBlock,
+		ErigonClient:   erigonClient,
+	}, nil
+}
+
+// Close closes the erigon RPC client
+func (mc *MigrationRPCService) Close() {
+	mc.ErigonClient.Close()
+}
+
+// shouldProxy determines if a request should be proxied based on block number
+func (mc *MigrationRPCService) shouldProxy(blockNumber uint64) bool {
+	return mc.MigrationBlock > 0 && blockNumber < mc.MigrationBlock
+}
+
+// MigrationBlockChainAPI wraps the standard BlockChainAPI to add migration routing
+type MigrationBlockChainAPI struct {
+	*ethapi.BlockChainAPI
+	config *MigrationRPCService
+}
+
+// NewMigrationBlockChainAPI creates a new migration-aware BlockChainAPI
+func NewMigrationBlockChainAPI(original *ethapi.BlockChainAPI, config *MigrationRPCService) *MigrationBlockChainAPI {
+	return &MigrationBlockChainAPI{
+		BlockChainAPI: original,
+		config:        config,
+	}
+}
+
+// eth_call
+// FORWARD
+func (api *MigrationBlockChainAPI) Call(ctx context.Context, args ethapi.TransactionArgs, blockNrOrHash *rpc.BlockNumberOrHash, overrides *override.StateOverride, blockOverrides *override.BlockOverrides) (hexutil.Bytes, error) {
+	if blockNr, ok := blockNrOrHash.Number(); ok && blockNr >= 0 {
+		if api.config.shouldProxy(uint64(blockNr)) {
+			var result hexutil.Bytes
+			err := api.config.ErigonClient.CallContext(ctx, &result, "eth_call", args, blockNrOrHash, overrides)
+			return result, err
+		}
+	}
+
+	localResult, err := api.BlockChainAPI.Call(ctx, args, blockNrOrHash, overrides, blockOverrides)
+	if err == nil && localResult != nil {
+		return localResult, nil
+	}
+
+	var result hexutil.Bytes
+	err = api.config.ErigonClient.CallContext(ctx, &result, "eth_call", args, blockNrOrHash, overrides)
+	return result, err
+}
```

Comment:
if given block > migrate_number, will go to line 100. But if line 100 failed, will go to 106. Maybe that is not our demand. Because we only want to handle the the block(> migrate_number) in local.

Reply:
Same issue at other functions below